File: 223 Rectangel Area.c
#include <stdio.h>
/*
223 Rectangle Area
https://leetcode.com/problems/rectangle-area/
(A,B) lower-left
(C,D) upper-right
*/
int dist(int x, int y){
    return x - y > 0 ? x - y : y - x;
}
/*
0. Only (A,B) inside EFGH: 
area -= (G - A) * (H - B);
1. Only (A,B), (A,D) inside EFGH
area -= (G - A) * (D - B);
2. Only (A,B), (C,B) inside EFGH
area -= (C - A) * (H - B);
3. ABCD inside EFGH
area -= (C - A) * (D - B);
*/
#define Compare(area,A,B,C,D,E,F,G,H)\
    if((A >= E && A <= G || A <= E && A >= G )&&(B >= F && B <= H || B <= F && B >= H)){\
        int ca = dist(C,A);\
        int ga = dist(G,A);\
        int db = dist(D,B);\
        int hb = dist(H,B);\
        return area - (ca < ga ? ca : ga) * (db < hb ? db: hb);\
    }

int computeArea(int A, int B, int C, int D, int E, int F, int G, int H)
{
    int area = (C - A)*(D - B) + (G - E )*(H - F);
    
    Compare(area,A,B,C,D,E,F,G,H);
    Compare(area,C,B,A,D,G,F,E,H);
    Compare(area,C,D,A,B,G,H,E,F);
    Compare(area,A,D,C,B,E,H,G,F);
    
    Compare(area,E,F,G,H,A,B,C,D);
    Compare(area,G,F,E,H,C,B,A,D);
    Compare(area,G,H,E,F,C,D,A,B);
    Compare(area,E,H,G,F,A,D,C,B);
    
    // non-overlap
    return area;
}
void test(void)
{
    int arr[] = {-2,-2,2,2,-1,-1,1,1};//16
    //int arr[] = {2,2,3,3,-2,-2,2,2};//17
    //int arr[] = {-2,-2,2,2,2,2,3,3};//17
    //int arr[] = {-2, -2, 2, 2, 1, -3, 3, -1}; // 19

    int r = computeArea(\
        arr[0],arr[1],arr[2],arr[3],\
        arr[4],arr[5],arr[6],arr[7]);
    
    printf("%d\n",r);
}
int main()
{
    test();
    return 0;
}

File: 223 Rectangel Area.py
'''
223. Rectangle Area
https://leetcode.com/problems/rectangle-area/

Find the total area covered by two rectilinear rectangles in a 2D plane.

Each rectangle is defined by its bottom left corner and top right corner as shown in the figure.

Rectangle Area

Example:

Input: A = -3, B = 0, C = 3, D = 4, E = 0, F = -1, G = 9, H = 2
Output: 45
'''
from typing import Tuple

class Solution:
    def overlappedInterval(self, i1: Tuple[int, int], i2: Tuple[int, int]) -> int:
        '''
        Return the overlapped length of 2 intervals
        '''
        return max(0, min(i1[1], i2[1]) - max(i1[0], i2[0]))
        
    def overlappedArea(self, A: int, B: int, C: int, D: int, E: int, F: int, G: int, H: int) -> int:
        return self.overlappedInterval((A, C), (E, G)) * self.overlappedInterval((B, D), (F, H))
        
    def rectangleArea(self, A: int, B: int, C: int, D: int) -> int:
        return abs((A-C) * (B-D))
        
    def computeArea(self, A: int, B: int, C: int, D: int, E: int, F: int, G: int, H: int) -> int:
        return self.rectangleArea(A,B,C,D) + self.rectangleArea(E,F,G,H) - self.overlappedArea(A,B,C,D,E,F,G,H)
